### ISSUE TITLE
Inconsistent publishing rate for sensors

### ISSUE THREAD

**Issue by alonsollorente** [Field-Robotics-Japan/UnitySensors#166]
Hi everyone! 

I hope you can help me with this problem that I'm having. I'm trying to broadcast sensory data in ROS2 and going through some difficulties. 

Thank you in advance for helping me out! I let you with a brief description of my problem.

**Description of the bug**
There is an inconsistency in the publishing frequency of sensor data between ROS2 and Unity when using ros_tcp_endpoint. Despite setting the publishing rate to 10Hz for one LiDAR sensor and 100Hz for an IMU, the actual publishing rate observed is significantly lower and inconsistent, being in the case of the IMU around 35Hz. The ```ros2 topic hz /imu``` output can be seen:
 
![image](https://github.com/user-attachments/assets/3dc680d9-c9bc-4d09-ad83-1cc3780a2443)

**To Reproduce**
- I'm using the set of sensors from [UnitySensor](https://github.com/Field-Robotics-Japan/UnitySensors), which uses ROS_TCP_Connector.
- Setting from IMU and  LiDAR:

![image](https://github.com/user-attachments/assets/efcd2258-a9f3-4dbc-ba64-fb0ee3b61510)
![image](https://github.com/user-attachments/assets/92aba001-51b1-44fa-99de-2f22656357ba)

**Environment set up:**
- Unity Version: 2023.2.20f1
- Ubuntu 22.04, ROS2 Humble

**Comment by RyodoTanaka** [comment]
@alonsollorente 
Thank you for updating the issue.
This issue is related to https://github.com/Field-Robotics-Japan/UnitySensors/issues/156 (It is written in Japanese, but discussing same topic)
The main reason of this problem comes from using Update() function.
To fix this, we should use FixUpdate() function.
https://github.com/Field-Robotics-Japan/UnitySensors/blob/master/Assets/UnitySensorsROS/Runtime/Scripts/Publishers/RosMsgPublisher.cs#L39
https://github.com/Field-Robotics-Japan/UnitySensors/blob/master/Assets/UnitySensors/Runtime/Scripts/Sensors/IMU/IMUSensor.cs#L51

But, this problem is already fixed in `develop` branch.
Thus, please try to use `develop` branch.
Now, merging PR from `develop` into `master` is under development (https://github.com/Field-Robotics-Japan/UnitySensors/pull/163)

**Comment by alonsollorente** [comment]
Hi!

Thank you for the fast answer. I checked that the Update() function in the LiDAR publisher case is the same in the ```develop```  and ```master```   branch. I don't know if maybe I'm approaching it wrong...

Furthermore, I tried the project in the ```develop```  branch, and is still inconsistent in frequency for both sensors.

**Comment by RyodoTanaka** [comment]
@alonsollorente 
>  the Update() function in the LiDAR publisher case is the same in the `develop`.

Yes, you are right. It is still not fixed in `develop` branch about LiDAR Publisher.

> I tried the project in the develop branch, and is still inconsistent in frequency for both sensors.

About this, I need to find the reason why especially for IMU sensors.
I think it takes a bit long time to fix. But if someone find the solution, comment it please.

**Comment by Bob-Eric** [comment]
Hi there!
I have tried these sensors in both ROS1 and ROS2 and they work fine.
However, you have to be aware of the **frame rate** in Unity. 
Currently it is not possible to publish sensors faster than the frame rate.

<img width="1280" alt="{F2949AC1-41AE-4560-B41F-48C1BF3B04B7}" src="https://github.com/user-attachments/assets/474198e9-89d7-4b43-98d9-ccbe191f8d73">

<img width="1280" alt="{8325D53A-8857-4213-BC92-33AA2162E3A1}" src="https://github.com/user-attachments/assets/e2a296de-33b9-474c-a3cd-173866a8b925">

**Comment by RyodoTanaka** [comment]
This Issue would related with following.
https://github.com/Field-Robotics-Japan/UnitySensors/pull/188/files#diff-5d75f813dd50b7d273b40a2fe62b8e907eb1bf7e23aac2d6fd35685f24272d34R54

**Comment by wangguohao-github** [comment]
Hi, i have fixed this problem. You could replace unitysensor.cs by the following.

`using UnityEngine;
// using Unity.Robotics.ROSTCPConnector;

namespace UnitySensors.Sensor
{
    public abstract class UnitySensor : MonoBehaviour
    {
        [SerializeField]
        private float _frequency = 10.0f;

        private float _time;
        private float _dt;
        private float _frequency_inv;
        private float _dt_step;

        public delegate void OnSensorUpdated();
        public OnSensorUpdated onSensorUpdated;

        public float sensorDeltaTime => _dt_step;


        public float time => _time;

        private void Awake()
        {
            _dt = 0.0f;
            _frequency_inv = 1.0f / _frequency;
            Init();
        }

        protected virtual void Update()
        {
            _dt += Time.deltaTime;

            while (_dt >= _frequency_inv)
            {
                _dt -= _frequency_inv;

                _time = Time.time;
                _dt_step = _frequency_inv;

                UpdateSensor();
            }
        }

        private void OnDestroy()
        {
            onSensorUpdated = null;
            OnSensorDestroy();
        }

        protected abstract void Init();
        protected abstract void UpdateSensor();
        protected abstract void OnSensorDestroy();
    }
}
`
